File: components/common/src/main/java/com/hotels/styx/server/HttpInterceptorContext.java
/*
  Copyright (C) 2013-2018 Expedia Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package com.hotels.styx.server;

import com.hotels.styx.api.HttpInterceptor;

import java.net.InetSocketAddress;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * ConcurrentHashMap backed implementation of HttpInterceptor.Context.
 */
public final class HttpInterceptorContext implements HttpInterceptor.Context {
    private final Map<String, Object> context = new ConcurrentHashMap<>();

    private final boolean secure;
    private final InetSocketAddress clientAddress;

    /**
     * Construct a new instance.
     *
     * @param secure true if the request was received via SSL
     * @param clientAddress address that request came from, or null if not-applicable
     */
    public HttpInterceptorContext(boolean secure, InetSocketAddress clientAddress) {
        this.secure = secure;
        this.clientAddress = clientAddress; // intentionally nullable
    }

    /**
     * Construct a new instance.
     *
     * @param clientAddress address that request came from, or null if not-applicable
     */
    public HttpInterceptorContext(InetSocketAddress clientAddress) {
        this(false, clientAddress);
    }

    /**
     * Construct a new instance.
     *
     * @param secure true if the request was received via SSL
     */
    public HttpInterceptorContext(boolean secure) {
        this(secure, null);
    }

    /**
     * Construct a new instance.
     */
    public HttpInterceptorContext() {
        this(false, null);
    }

    // TODO deprecate
    public static HttpInterceptor.Context create() {
        return new HttpInterceptorContext();
    }

    @Override
    public void add(String key, Object value) {
        context.put(key, value);
    }

    @Override
    public <T> T get(String key, Class<T> clazz) {
        return (T) context.get(key);
    }

    @Override
    public boolean isSecure() {
        return secure;
    }

    @Override
    public Optional<InetSocketAddress> clientAddress() {
        return Optional.ofNullable(clientAddress);
    }
}

File: components/common/src/main/kotlin/com/hotels/styx/server/HttpInterceptorContext.kt
/*
  Copyright (C) 2013-2024 Expedia Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package com.hotels.styx.server

import com.hotels.styx.api.HttpInterceptor
import com.hotels.styx.metrics.ContextualTimers
import com.hotels.styx.metrics.TimeMeasurable
import java.net.InetSocketAddress
import java.util.Optional
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executor

/**
 * ConcurrentHashMap backed implementation of HttpInterceptor.Context.
 *
 * @param secure true if the request was received via SSL
 * @param clientAddress address that request came from, or null if not-applicable
 */
class HttpInterceptorContext(
    private val secure: Boolean,
    private val clientAddress: InetSocketAddress?,
    private val executor: Executor?,
    override val timers: ContextualTimers? = null,
) : HttpInterceptor.Context, TimeMeasurable {
    // This may seem redundant but it allows Executor to be a lambda without needing to set `timers`.
    constructor(secure: Boolean, clientAddress: InetSocketAddress?, executor: Executor?) : this(secure, clientAddress, executor, null)

    private val context = ConcurrentHashMap<String, Any>()

    override fun add(
        key: String,
        value: Any,
    ) {
        context[key] = value
    }

    @Deprecated("Deprecated in Java")
    override fun <T> get(
        key: String,
        clazz: Class<T>,
    ): T = context[key] as T

    override fun isSecure() = secure

    override fun clientAddress() = Optional.ofNullable(clientAddress)

    override fun executor(): Executor? = executor

    override fun clear() = context.clear()

    companion object {
        // Visible for testing
        @Deprecated("use the constructor instead")
        @JvmStatic
        fun create(): HttpInterceptor.Context = HttpInterceptorContext(false, null) { it.run() }
    }
}
